internal/controller/metricsaggregate.go

```diff
@@ -0,0 +1,185 @@
+	logger := log.FromContext(ctx)
+
+	apiGroups, err := r.DiscoveryClient.ServerGroups()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	metricsAPIAvailable := SupportedMetricsAPIVersionAvailable(apiGroups)
+	if !metricsAPIAvailable {
+		logger.V(5).Info("metrics api not available")
+		return requeue(time.Minute*5, jitter), nil
+	}
+
+	// Read resource from Kubernetes cluster.
+	metrics := &v1alpha1.MetricsAggregate{}
+	if err := r.Get(ctx, req.NamespacedName, metrics); err != nil {
+		logger.Error(err, "unable to fetch MetricsAggregate")
+		return ctrl.Result{}, k8sClient.IgnoreNotFound(err)
+	}
+
+	logger.Info("reconciling MetricsAggregate", "namespace", metrics.Namespace, "name", metrics.Name)
+	utils.MarkCondition(metrics.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, "")
+
+	scope, err := NewDefaultScope(ctx, r.Client, metrics)
+	if err != nil {
+		logger.Error(err, "failed to create scope")
+		utils.MarkCondition(metrics.SetCondition, v1alpha1.ReadyConditionType, metav1.ConditionFalse, v1alpha1.ReadyConditionReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Always patch object when exiting this function, so we can persist any object changes.
+	defer func() {
+		if err := scope.PatchObject(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	if !metrics.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	nodeList := &corev1.NodeList{}
+	if err := r.List(ctx, nodeList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	availableResources := make(map[string]corev1.ResourceList)
+	for _, n := range nodeList.Items {
+		availableResources[n.Name] = n.Status.Allocatable
+	}
+
+	nodeDeploymentNodesMetrics := make([]v1beta1.NodeMetrics, 0)
+	allNodeMetricsList := &v1beta1.NodeMetricsList{}
+	if err := r.List(ctx, allNodeMetricsList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	for _, m := range allNodeMetricsList.Items {
+		if _, ok := availableResources[m.Name]; ok {
+			nodeDeploymentNodesMetrics = append(nodeDeploymentNodesMetrics, m)
+		}
+	}
+
+	nodeMetrics, err := ConvertNodeMetrics(nodeDeploymentNodesMetrics, availableResources)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// save metrics
+	metrics.Spec.Nodes = len(nodeList.Items)
+	for _, nm := range nodeMetrics {
+		metrics.Spec.CPUAvailableMillicores += nm.CPUAvailableMillicores
+		metrics.Spec.CPUTotalMillicores += nm.CPUTotalMillicores
+		metrics.Spec.MemoryAvailableBytes += nm.MemoryAvailableBytes
+		metrics.Spec.MemoryTotalBytes += nm.MemoryTotalBytes
+	}
+
+	fraction := float64(metrics.Spec.CPUTotalMillicores) / float64(metrics.Spec.CPUAvailableMillicores) * 100
+	metrics.Spec.CPUUsedPercentage = int64(fraction)
+	fraction = float64(metrics.Spec.MemoryTotalBytes) / float64(metrics.Spec.MemoryAvailableBytes) * 100
+	metrics.Spec.MemoryUsedPercentage = int64(fraction)
+
+	return ctrl.Result{}, reterr
```

Comment:
shouldn't you be requeueing to update on an interval

Reply:
done